File: bindings/nodejs/examples/batch.ts
/**
 * Batch processing example for Pyjamaz Node.js bindings (TypeScript)
 */

import * as pyjamaz from '../src/index';
import * as fs from 'fs';
import * as path from 'path';
import { promisify } from 'util';

const readdir = promisify(fs.readdir);
const stat = promisify(fs.stat);

interface ProcessResult {
  success: boolean;
  input: string;
  output?: string;
  originalSize?: number;
  optimizedSize?: number;
  reduction?: number;
  format?: string;
  quality?: number;
  error?: string;
}

/**
 * Optimize a single image
 */
async function optimizeSingleImage(
  inputPath: string,
  outputDir: string,
  maxBytes: number
): Promise<ProcessResult> {
  try {
    const result = await pyjamaz.optimizeImage(inputPath, {
      maxBytes,
      metric: 'ssimulacra2',
      cacheEnabled: true,
    });

    if (result.passed) {
      const inputName = path.basename(inputPath, path.extname(inputPath));
      const outputPath = path.join(outputDir, `${inputName}_optimized.${result.format}`);
      await result.save(outputPath);

      const originalSize = (await stat(inputPath)).size;
      const reduction = ((1 - result.size / originalSize) * 100);

      return {
        success: true,
        input: path.basename(inputPath),
        output: path.basename(outputPath),
        originalSize,
        optimizedSize: result.size,
        reduction,
        format: result.format,
        quality: result.diffValue,
      };
    } else {
      return {
        success: false,
        input: path.basename(inputPath),
        error: result.errorMessage,
      };
    }
  } catch (error) {
    return {
      success: false,
      input: path.basename(inputPath),
      error: error instanceof Error ? error.message : String(error),
    };
  }
}

/**
 * Process images in batches with parallelism
 */
async function processBatch(
  images: string[],
  outputDir: string,
  maxBytes: number,
  maxParallel: number
): Promise<ProcessResult[]> {
  const results: ProcessResult[] = [];
  const pending: Promise<ProcessResult>[] = [];

  for (const imagePath of images) {
    // Add to pending promises
    const promise = optimizeSingleImage(imagePath, outputDir, maxBytes);
    pending.push(promise);

    // When we reach max parallel, wait for one to complete
    if (pending.length >= maxParallel) {
      const result = await Promise.race(pending);
      results.push(result);

      // Remove completed promise
      const index = pending.findIndex(async (p) => (await p) === result);
      if (index !== -1) {
        pending.splice(index, 1);
      }
    }
  }

  // Wait for remaining promises
  const remaining = await Promise.all(pending);
  results.push(...remaining);

  return results;
}

/**
 * Main function
 */
async function main() {
  // Configuration
  const inputDir = 'images';
  const outputDir = 'optimized';
  const maxBytes = 100_000; // 100KB target
  const maxParallel = 4; // Parallel workers

  // Create output directory
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  // Find all images
  const imageExtensions = new Set(['.jpg', '.jpeg', '.png', '.webp']);
  const files = await readdir(inputDir);
  const images = files
    .filter((file) => imageExtensions.has(path.extname(file).toLowerCase()))
    .map((file) => path.join(inputDir, file));

  console.log(`Found ${images.length} images to optimize`);
  console.log(`Target size: ${maxBytes.toLocaleString()} bytes`);
  console.log(`Using ${maxParallel} parallel workers`);
  console.log();

  // Process images
  const startTime = Date.now();
  const results: ProcessResult[] = [];

  // Process with progress updates
  for (const imagePath of images) {
    const result = await optimizeSingleImage(imagePath, outputDir, maxBytes);
    results.push(result);

    if (result.success) {
      console.log(
        `✓ ${result.input}: ${result.originalSize!.toLocaleString()} → ` +
        `${result.optimizedSize!.toLocaleString()} bytes (${result.reduction!.toFixed(1)}% reduction)`
      );
    } else {
      console.log(`✗ ${result.input}: ${result.error}`);
    }
  }

  const elapsed = (Date.now() - startTime) / 1000;

  // Print summary
  const successful = results.filter((r) => r.success);
  const failed = results.filter((r) => !r.success);

  console.log(`\n${'='.repeat(60)}`);
  console.log(`Processed ${images.length} images in ${elapsed.toFixed(2)}s`);
  console.log(`Success: ${successful.length}, Failed: ${failed.length}`);

  if (successful.length > 0) {
    const totalOriginal = successful.reduce((sum, r) => sum + (r.originalSize || 0), 0);
    const totalOptimized = successful.reduce((sum, r) => sum + (r.optimizedSize || 0), 0);
    const totalReduction = ((1 - totalOptimized / totalOriginal) * 100);

    console.log(`\nTotal size: ${totalOriginal.toLocaleString()} → ${totalOptimized.toLocaleString()} bytes`);
    console.log(`Total reduction: ${totalReduction.toFixed(1)}%`);
    console.log(`Average per image: ${(elapsed / images.length).toFixed(3)}s`);
  }
}

// Run
main().catch(console.error);

File: bindings/python/examples/batch.py
"""
Batch processing example for Pyjamaz.
"""

import pyjamaz
from pathlib import Path
from concurrent.futures import ThreadPoolExecutor, as_completed
import time

def optimize_single_image(input_path: Path, output_dir: Path, max_bytes: int):
    """Optimize a single image."""
    try:
        result = pyjamaz.optimize_image(
            str(input_path),
            max_bytes=max_bytes,
            metric='ssimulacra2',
            cache_enabled=True,  # Use cache for speed
        )

        if result.passed:
            output_path = output_dir / f"{input_path.stem}_optimized.{result.format}"
            result.save(output_path)

            original_size = input_path.stat().st_size
            reduction = (1 - result.size / original_size) * 100

            return {
                'success': True,
                'input': input_path.name,
                'output': output_path.name,
                'original_size': original_size,
                'optimized_size': result.size,
                'reduction': reduction,
                'format': result.format,
                'quality': result.diff_value,
            }
        else:
            return {
                'success': False,
                'input': input_path.name,
                'error': result.error_message,
            }
    except Exception as e:
        return {
            'success': False,
            'input': input_path.name,
            'error': str(e),
        }

def main():
    # Configuration
    input_dir = Path('images')
    output_dir = Path('optimized')
    max_bytes = 100_000  # 100KB target
    max_workers = 4  # Parallel workers

    # Create output directory
    output_dir.mkdir(exist_ok=True)

    # Find all images
    image_extensions = {'.jpg', '.jpeg', '.png', '.webp'}
    images = [f for f in input_dir.iterdir() if f.suffix.lower() in image_extensions]

    print(f"Found {len(images)} images to optimize")
    print(f"Target size: {max_bytes:,} bytes")
    print(f"Using {max_workers} parallel workers")
    print()

    # Process images in parallel
    start_time = time.time()
    results = []

    with ThreadPoolExecutor(max_workers=max_workers) as executor:
        # Submit all tasks
        future_to_image = {
            executor.submit(optimize_single_image, img, output_dir, max_bytes): img
            for img in images
        }

        # Collect results as they complete
        for future in as_completed(future_to_image):
            result = future.result()
            results.append(result)

            if result['success']:
                print(f"✓ {result['input']}: "
                      f"{result['original_size']:,} → {result['optimized_size']:,} bytes "
                      f"({result['reduction']:.1f}% reduction)")
            else:
                print(f"✗ {result['input']}: {result['error']}")

    elapsed = time.time() - start_time

    # Print summary
    successful = [r for r in results if r['success']]
    failed = [r for r in results if not r['success']]

    print(f"\n{'='*60}")
    print(f"Processed {len(images)} images in {elapsed:.2f}s")
    print(f"Success: {len(successful)}, Failed: {len(failed)}")

    if successful:
        total_original = sum(r['original_size'] for r in successful)
        total_optimized = sum(r['optimized_size'] for r in successful)
        total_reduction = (1 - total_optimized / total_original) * 100

        print(f"\nTotal size: {total_original:,} → {total_optimized:,} bytes")
        print(f"Total reduction: {total_reduction:.1f}%")
        print(f"Average per image: {elapsed/len(images):.3f}s")

if __name__ == '__main__':
    main()
